File: real_benchmarks/data/real_c_rust_pairs/uninitialized_memory.c
/*
 * Uninitialized memory divergence — inspired by CVE-2019-15166 (tcpdump)
 * and CVE-2017-12172 (PostgreSQL uninit read).
 *
 * C11 §6.3.2.1/2: reading an uninitialized variable of automatic storage
 * duration that could have been register-qualified is UB.
 * Rust: all variables must be initialized before use (compiler-enforced).
 */
#include <stdint.h>
#include <string.h>

/* Classic: uninitialized variable on error path. */
int parse_header(const uint8_t *buf, int len) {
    int version;
    int flags;
    if (len >= 2) {
        version = buf[0];
        flags = buf[1];
    }
    /* If len < 2, version and flags are uninitialized — UB to read them */
    return version + flags;
}

/* Partial struct initialization (common real-world bug). */
struct Config {
    int mode;
    int timeout;
    int retries;
    int verbose;
};

int get_total_config(int has_custom) {
    struct Config cfg;
    cfg.mode = 1;
    if (has_custom) {
        cfg.timeout = 30;
        cfg.retries = 3;
        cfg.verbose = 0;
    }
    /* If !has_custom, timeout/retries/verbose are uninit — UB */
    return cfg.mode + cfg.timeout + cfg.retries + cfg.verbose;
}

/* Stack buffer: not zeroed by default in C. */
int sum_uninitialized_array(int n) {
    int arr[64];  /* uninitialized stack array */
    /* Forgot to initialize — reading arr[i] is UB */
    int sum = 0;
    for (int i = 0; i < n && i < 64; i++) {
        sum += arr[i];
    }
    return sum;
}

/* Conditional initialization: optimizer may assume init. */
int conditional_init(int selector, int a, int b) {
    int result;
    if (selector > 0)
        result = a;
    else if (selector < 0)
        result = b;
    /* if selector == 0, result is uninitialized — UB */
    return result;
}

File: real_benchmarks/data/real_c_rust_pairs/uninitialized_memory.rs
/*
 * Rust translation of uninitialized memory patterns.
 * Rust's ownership system requires all variables be initialized before use.
 * The compiler refuses to compile if any path leaves a variable uninitialized.
 */

/// Parse header: Rust requires explicit initialization on all paths.
pub fn parse_header(buf: &[u8], len: i32) -> i32 {
    let version: i32;
    let flags: i32;
    if len >= 2 {
        version = buf[0] as i32;
        flags = buf[1] as i32;
    } else {
        version = 0;  // Rust forces initialization on else path
        flags = 0;
    }
    version + flags
}

/// Partial struct init: Rust requires all fields be initialized.
pub struct Config {
    pub mode: i32,
    pub timeout: i32,
    pub retries: i32,
    pub verbose: i32,
}

pub fn get_total_config(has_custom: bool) -> i32 {
    let cfg = if has_custom {
        Config { mode: 1, timeout: 30, retries: 3, verbose: 0 }
    } else {
        Config { mode: 1, timeout: 0, retries: 0, verbose: 0 }  // must init all
    };
    cfg.mode + cfg.timeout + cfg.retries + cfg.verbose
}

/// Stack array: Rust arrays must be initialized.
pub fn sum_uninitialized_array(n: i32) -> i32 {
    let arr = [0i32; 64];  // must initialize
    let mut sum = 0i32;
    let limit = n.min(64) as usize;
    for i in 0..limit {
        sum = sum.wrapping_add(arr[i]);
    }
    sum
}

/// Conditional init: Rust compiler rejects if any path leaves uninit.
pub fn conditional_init(selector: i32, a: i32, b: i32) -> i32 {
    let result: i32;
    if selector > 0 {
        result = a;
    } else if selector < 0 {
        result = b;
    } else {
        result = 0;  // Rust forces a default
    }
    result
}
